### ISSUE TITLE
Support dynamic wallpaper

### ISSUE THREAD

**Issue by astro-ray** [Wartybix/GNOME-Auto-Accent-Colour#20]
![Screenshot From 2024-12-01 12-30-11](https://github.com/user-attachments/assets/d6a6e481-6530-47ac-b7b5-536bc43bb2cc)

I have dynamic wallpaper set and the extension doesn't parse xml files.

Is it in the scope of this extension to add this feature?

**Comment by Wartybix** [comment]
Yes, I would like to add XML support to my extension. After all, wallpapers that change automatically are the main use case for Auto Accent Colour.

When I was testing this extension, I had to purposefully block it from running on XML files (hence why you see that notification), otherwise it would crash. The extension fetches (and re-runs on updates of) the image pointed to by the `/org/gnome/desktop/background/picture-uri` or `picture-uri-dark` gsettings key, which it then parses colour data from and sets an accent colour accordingly. Normally, the value of one of these keys is the URI of a .jpg, .png., .svg, etc. file. However, when this value is an .xml file, the extension doesn't know what to do with this, as the .xml file isn't an image file itself, just a list of paths to images with duration/transition times to essentially make up a slide-show. So no colour data can be parsed from this file directly.

What I would need to do is find some way for the extension to calculate what the 'current' picture is based on timings in the .xml file, as well as schedule when to re-run the script based on those same timings, so that a new accent colour can be promptly applied when the dynamic wallpaper advances onto the next image. I don't know how to do this yet. When I have time, I might have a look around at `gnome-control-center`'s source code (and that of its Ubuntu counterpart) to see if I can spot the algorithm to get the 'current' picture of an .xml file, and try to implement the same functionality into my extension. If you have any good ideas, please feel free to submit a pull request too.

In the meantime, some nice dynamic wallpaper software I can recommend is the Bing Wallpaper GNOME extension, or the Variety app. These apps can automatically set a new 'image of the day' (from sources like Bing or NASA's daily image) as your background, or cycle/shuffle through pictures in a favourites folder. And they do this by setting the `picture-uri`/`picture-uri-dark` key value directly to an image file, so Auto Accent Colour works with them. Doesn't always meet the functionality of XML files (especially those time of day ones), but still, quite nice alternatives.

**Comment by yannvillellas** [comment]
I would be interested too!
